improve the tests with real KF profile namespaces

kunal-511 opened the pull request:
# Pull Request Template for Kubeflow Manifests

## ✏️ Summary of Changes
> Added pending tests to end-to-end integration tests
> Updated the katib and training operator to test in real KF profile namespace
> Added missing things to the volume web application test


## 🐛 Related Issues
> Fix #3054

## ✅ Contributor Checklist
  - [ ] I have tested these changes with kustomize. See [Installation Prerequisites](https://github.com/kubeflow/manifests#prerequisites).
  - [x] All commits are [_signed-off_](https://github.com/kubeflow/community/tree/master/dco-signoff-hook#signing-off-commits) to satisfy the DCO check.

---     
 
> You can join the CNCF Slack and access our meetings at the [Kubeflow Community](https://www.kubeflow.org/docs/about/community/) website. Our channel on the CNCF Slack is here [**#kubeflow-platform**](https://app.slack.com/client/T08PSQ7BQ/C073W572LA2).

juliusvonkohout commented on .github/workflows/katib_test.yaml:
```diff
@@ -22,35 +25,56 @@ jobs:
-        ./tests/gh-actions/install_katib.sh
+      run: ./tests/gh-actions/install_katib.sh
+
+    - name: Install Training Operator
```
why do you install training-operator instead of Katib in the katib test ?

kunal-511 replied:
because i thought Katib relies on the training operator for some of its experiment functionality ?

juliusvonkohout replied:
I think it is a separate thing.

juliusvonkohout commented on .github/workflows/training_operator_test.yaml:
```diff
@@ -26,31 +30,61 @@ jobs:
+      run: ./tests/gh-actions/test_training_operator.sh "${KF_PROFILE}"
+
+    - name: Test with Authorized Token
+      run: kubectl get pytorchjobs -n $KF_PROFILE --token="$(kubectl -n $KF_PROFILE create token default-editor --audience=training.kubeflow.org)"
```
Why do you add --audience=training.kubeflow.org?

juliusvonkohout replied:
Does it improve security?

juliusvonkohout commented on tests/gh-actions/test_volumes_web_app.sh:
```diff
@@ -0,0 +1,104 @@
+cat <<EOF | kubectl apply -f -
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: test-pv
+spec:
+  capacity:
+    storage: 1Gi
+  accessModes:
+    - ReadWriteOnce
+  persistentVolumeReclaimPolicy: Delete
+  volumeMode: Filesystem
+  local:
+    path: /tmp/data
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: kubernetes.io/hostname
+          operator: In
+          values:
+          - kind-control-plane
+EOF
+
+kubectl exec -it kind-control-plane -- mkdir -p /tmp/data
+
+cat <<EOF | kubectl apply -f -
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test-pvc
+  namespace: $KF_PROFILE
+spec:
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
+  storageClassName: standard
+EOF
```
please remove this and use the Volumes API instead. just the same as using the ui with a webbrowser. you can use the debugging tools in your browser to get the curl commands.

juliusvonkohout commented on tests/gh-actions/test_training_operator.sh:
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+set -euxo pipefail
```
set -euxo

juliusvonkohout commented on tests/gh-actions/install_training_operator.sh:
```diff
@@ -1,12 +1,23 @@
+
+kubectl wait --for=condition=Available deployment/training-operator -n kubeflow --timeout=180s
+
+for i in {1..60}; do
+  if kubectl get crd pytorchjobs.kubeflow.org >/dev/null 2>&1 && 
+     kubectl api-resources | grep -q "pytorchjobs.*kubeflow"; then
+    break
+  fi
+  
+  [ $i -eq 60 ] && exit 1
+  sleep 2
+done
```
is this necessary? If the job finishes it should be fine
```suggestion
```

kunal-511 replied:
this was added for some debugging i will eventually remove it

kunal-511 commented:
@juliusvonkohout can you please review this

juliusvonkohout commented on tests/gh-actions/test_volumes_web_app.sh:
```diff
@@ -0,0 +1,87 @@
+CSRF_COOKIE=$(curl -s -c - "localhost:8080/volumes/" | grep XSRF-TOKEN | cut -f 7)
+CSRF_HEADER=${CSRF_COOKIE}
+
+SC_NAME="standard"
```
storage_class_name

juliusvonkohout commented on .github/workflows/full_kubeflow_integration_test.yaml:
```diff
@@ -111,7 +114,13 @@ jobs:
-        python3 tests/gh-actions/pipeline_test.py test_unauthorized_access "$(kubectl -n default create token default)" "${KF_PROFILE}"
+        kubectl create namespace test-unauthorized
+        kubectl create serviceaccount test-unauthorized -n test-unauthorized
+        UNAUTH_TOKEN=$(kubectl -n test-unauthorized create token test-unauthorized)
```
UNAUTHORIZED_TOKEN

juliusvonkohout commented on .github/workflows/full_kubeflow_integration_test.yaml:
```diff
@@ -69,6 +69,9 @@ jobs:
     - name: Install PodDefaults CRD
       run: kubectl get crd poddefaults.kubeflow.org || kubectl apply -f https://raw.githubusercontent.com/kubeflow/kubeflow/master/components/admission-webhook/manifests/base/crd.yaml
 
+    - name: Install Volumes Web App
```
Install Volumes Web Application

juliusvonkohout commented on .github/workflows/full_kubeflow_integration_test.yaml:
```diff
@@ -139,8 +148,14 @@ jobs:
+        sleep 30
+        kubectl get inferenceservice -n $KF_PROFILE
+        kubectl wait --for=condition=Ready inferenceservice.serving.kserve.io/sklearn-iris -n $KF_PROFILE --timeout=300s
 
```
```suggestion
# TODO the individual KServe tests is currently being restructured. Afterwards we can also test inferencing

```

juliusvonkohout commented on tests/gh-actions/install_volumes_web_app.sh:
```diff
@@ -0,0 +1,9 @@
+kustomize build overlays/istio | kubectl apply -f -
+cd ../../../
+
+sleep 5 
```
I think you need to wait for the deployment here

juliusvonkohout commented on tests/gh-actions/test_training_operator.sh:
```diff
@@ -0,0 +1,15 @@
+sed 's/name: pytorch-simple/name: pytorch-simple\n  namespace: '"$KF_PROFILE"'/g' > /tmp/pytorch-job.yaml
+kubectl apply -f /tmp/pytorch-job.yaml
+
+sleep 90
```
I think we can do better than sleep

juliusvonkohout commented on tests/gh-actions/test_volumes_web_app.sh:
```diff
@@ -0,0 +1,87 @@
+  "localhost:8080/volumes/api/namespaces/${KF_PROFILE}/pvcs/test-pvc" \
+  -H "Authorization: Bearer ${TOKEN}" \
+  -H "X-XSRF-TOKEN: ${CSRF_HEADER}" \
+  -b "XSRF-TOKEN=${CSRF_COOKIE}" > /dev/null
```
how do you check the result ?

juliusvonkohout replied:
What about printing the response and checking for http code  200?

juliusvonkohout replied:
Done

juliusvonkohout commented on tests/gh-actions/test_volumes_web_app.sh:
```diff
@@ -0,0 +1,87 @@
+  "localhost:8080/volumes/api/namespaces/${KF_PROFILE}/pvcs/test-pvc" \
+  -H "Authorization: Bearer ${UNAUTH_TOKEN}" \
+  -H "X-XSRF-TOKEN: ${CSRF_HEADER}" \
+  -b "XSRF-TOKEN=${CSRF_COOKIE}" > /dev/null
```
how do you check the success ?

juliusvonkohout replied:
What about printing the response and checking for http code  200?

juliusvonkohout commented on tests/gh-actions/test_volumes_web_app.sh:
```diff
@@ -0,0 +1,87 @@
+      },
+      \"storageClassName\": \"${SC_NAME}\"
+    }
+  }" > /dev/null
```
how do you check the success ?

juliusvonkohout replied:
What about printing the response and checking for http code  200?

juliusvonkohout commented on .github/workflows/training_operator_test.yaml:
```diff
@@ -5,13 +5,17 @@ on:
-    - tests/gh-actions/kf-objects/tfjob.yaml
-    - tests/gh-actions/install_istio.sh
+    - tests/gh-actions/kf-objects/training_operator_job.yaml
+    - tests/gh-actions/install_istio-cni.sh
```
We need to make sure than that we use istio-cni everywhere

juliusvonkohout commented:
> @juliusvonkohout can you please review this

Looks promising, thank you. i created comments.

juliusvonkohout commented on .github/workflows/full_kubeflow_integration_test.yaml:
```diff
@@ -139,8 +148,15 @@ jobs:
       run: |
-        sed 's/namespace: .*/namespace: '"$KF_PROFILE"'/g' tests/gh-actions/kf-objects/training_operator_job.yaml | kubectl apply -f -
+        sed "s/PROFILE_NAMESPACE/$KF_PROFILE/g" tests/gh-actions/kf-objects/kserve_test.yaml | kubectl apply -f -
+        sleep 30
```
I think we can do better than sleep here in a follow up PR. Inferenceservices have a readiness field.

juliusvonkohout commented on .github/workflows/katib_test.yaml:
```diff
@@ -22,35 +25,56 @@ jobs:
-        kubectl wait --for=condition=Running experiments.kubeflow.org -n kubeflow-user --all --timeout 300s
 
-        echo "Waiting for all Trials to become Succeeded..."
-        kubectl wait --for=condition=Succeeded trials.kubeflow.org -n kubeflow-user --all --timeout 600s
```
Where is the trials test now? I do not see it anymore?

juliusvonkohout commented on .github/workflows/training_operator_test.yaml:
```diff
@@ -26,31 +30,65 @@ jobs:
+      run: |
+        kubectl create namespace test-unauthorized
+        kubectl create serviceaccount test-unauthorized -n test-unauthorized
+        UNAUTH_TOKEN=$(kubectl -n test-unauthorized create token test-unauthorized)
```
Please do not use UNAUTH. Use UNAUTHORIZED everywhere

juliusvonkohout commented on .github/workflows/katib_test.yaml:
```diff
@@ -22,35 +25,56 @@ jobs:
-        echo "Waiting for all Trials to become Succeeded..."
-        kubectl wait --for=condition=Succeeded trials.kubeflow.org -n kubeflow-user --all --timeout 600s
+    - name: Test Authorized Katib Access
+      run: kubectl get experiments.kubeflow.org -n $KF_PROFILE --token="$(kubectl -n $KF_PROFILE create token default-editor --audience=katib.kubeflow.org)"
```
Why do you set the audience?

juliusvonkohout commented:
Looks good. Thank you. If you address the remaining comments we can merge it. The follow up PR would then to replace most of the single-step instructions in the root readme with just calling the install scripts. For example the simple single line install in https://github.com/kubeflow/manifests/blob/master/README.md#kubeflow-pipelines can be replaced with calling the install pipelines script

kunal-511 commented:
> Looks good. Thank you. If you address the remaining comments we can merge it. The next follow up PR would then be to replace most of the single-step instructions in the root readme with just calling the install scripts. For example the simple single line install in https://github.com/kubeflow/manifests/blob/master/README.md#kubeflow-pipelines can be replaced with calling the install pipelines script

i have addressed the comments

juliusvonkohout commented on .github/workflows/katib_test.yaml:
Maybe you have to disable the istio injection for Katib resources. Please check whether that is the reason for failure.

juliusvonkohout commented on .github/workflows/katib_test.yaml:
```diff
@@ -59,6 +59,11 @@ jobs:
+    - name: Disable Istio injection for Katib trials
+      run: |
+        kubectl label namespace $KF_PROFILE istio-injection- --overwrite
+        kubectl label namespace $KF_PROFILE sidecar.istio.io/inject=false --overwrite
```
No, that also blocks it for all other thest, please revert immideately.

juliusvonkohout commented on .github/workflows/katib_test.yaml:
```diff
@@ -59,6 +59,11 @@ jobs:
     - name: Port-forward the istio-ingress gateway
       run: ./tests/gh-actions/port_forward_gateway.sh
 
+    - name: Disable Istio injection for Katib trials
```
No, that also blocks it for all other tests, please revert immediately. Only for the Katib test

juliusvonkohout replied:
The best way would be to disable the injection cleanly in the katib objects directly

juliusvonkohout commented:
The best way would be to disable the injection cleanly in the katib objects directly . Do not disable istio for all tests in the profile namespace.

juliusvonkohout commented:
> > The best way would be to disable the injection cleanly in the katib objects directly . Do not disable istio for all tests in the profile namespace.
> 
> Is modifying the katib_test.yaml in the katib Test (gha) works ?

Yes, with .github/workflows/katib_test.yaml you can experiment.

juliusvonkohout commented on .github/workflows/katib_test.yaml:
```diff
@@ -61,7 +61,11 @@ jobs:
-        sed "s/kubeflow-user/$KF_PROFILE/g" tests/gh-actions/kf-objects/katib_test.yaml | kubectl apply -f -
+        sed "s/kubeflow-user/$KF_PROFILE/g" tests/gh-actions/kf-objects/katib_test.yaml > /tmp/katib_test_modified.yaml
+
+        sed -i '/spec:$/{n;s/template:/template:\n          metadata:\n            annotations:\n              sidecar.istio.io\/inject: "false"/}' /tmp/katib_test_modified.yaml
```
@kunal-511 please modify tests/gh-actions/kf-objects/katib_test.yaml directly instead of using sed.

juliusvonkohout replied:
please also remove other seds if possible.

kunal-511 replied:
okay

juliusvonkohout commented on tests/gh-actions/kf-objects/katib_test.yaml:
Really just go for namespace: kubeflow-user-example-com and remove sed

juliusvonkohout commented on tests/gh-actions/kf-objects/kserve_test.yaml:
Really just go for namespace: kubeflow-user-example-com and remove sed

juliusvonkohout commented:
Amazing
/lgtm
/approve
